runtime/lua/vim/img/_image.lua

```diff
@@ -0,0 +1,144 @@
+---@class vim.img.Image
+---@field name string|nil name of the image if loaded from disk
+---@field data string|nil base64 encoded data
+local M = {}
+M.__index = M
+
+---Creates a new image instance.
+---@param opts? {data?:string, filename?:string}
+---@return vim.img.Image
+function M:new(opts)
+  opts = opts or {}
+
+  local instance = {}
+  setmetatable(instance, M)
+
+  instance.data = opts.data
+  if not instance.data and opts.filename then
+    instance:load_from_file(opts.filename)
+  end
+
+  return instance
+end
+
+---Returns true if the image is loaded into memory.
+---@return boolean
+function M:is_loaded()
+  return self.data ~= nil
+end
+
+---Returns the size of the base64 encoded image.
+---@return integer
+function M:size()
+  return string.len(self.data or '')
+end
+
+---Iterates over the chunks of the image, invoking `f` per chunk.
+---@param f fun(chunk:string, pos:integer, has_more:boolean)
+---@param opts? {size?:integer}
+function M:for_each_chunk(f, opts)
```

Comment:
`:help dev-patterns`

```suggestion
function M:for_each_chunk(opts, f)
```

Reply:
Can we return an iterator instead? That is more idiomatic, and composes with `vim.iter`, `ipairs()`, etc.

Reply:
@justinmk yeah, I can do that. Was unaware of neovim having an iterator, so it'll be nice to get a feel for how that works.

Reply:
iterators are a Lua concept, and can be as simple as returning a function (closure) which, when called, returns the next item.

Reply:
`:h for-in`

Reply:
Updated to be `Image:chunks()` which returns the result of `vim.iter()` on a function iterator. Assuming that should be good enough?